### PR TITLE
Release Notes Feature

### PR DESCRIPTION
**Pull request by srevinsaju**

### CODE REVIEW

File: pyuploadtool/releases_hosting_provider/github_releases.py [hunk 1]
```diff
@@ -77,6 +78,13 @@ def create_release(self, metadata: ReleaseMetadata, artifacts):
 
         message = f"Build log: {metadata.build_log_url}"
 
+        if os.getenv("GENERATE_CHANGELOG", "").lower() == "true":
+            github_chglog = GitHubChangelogGenerator(
```
Comment (TheAssassin) [inline pyuploadtool/releases_hosting_provider/github_releases.py:82]: Not a fan of how this is integrated, as it mixes the "metadata gathering" with a use of metadata.

Changelogs are actually a specific type of metadata. You should probably extend the metadata class with a new property. Then, the GitHub actions builder can generate the metadata and fill it into the metadata object. Finally, the GitHub releases hosting provider can read those data (if available) and update the description.

File: pyuploadtool/releases_hosting_provider/github_releases.py [hunk 2]
```diff
@@ -77,6 +78,13 @@ def create_release(self, metadata: ReleaseMetadata, artifacts):
...
+                metadata=metadata,
+                github_token=os.environ["GITHUB_TOKEN"]
+            )
+            metadata.release_description = github_chglog.get_changelog()
```
[the remark this answers is not in the document]
Comment (TheAssassin) [reply pyuploadtool/releases_hosting_provider/github_releases.py:86]: Also, I think it makes sense to separate the changelog data type and the generation of suitable Markdown.

**Review by TheAssassin — CHANGES_REQUESTED** [review]

File: pyuploadtool/changelog/changelog.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+
+import os
+
+
+# order them by priority
+# https://cheatography.com/albelop/cheat-sheets/conventional-commits/
+CONVENTIONAL_COMMIT_SPEC = {
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:7]: There's no reason to process any kind of "commit prefix". I don't use them, hardly anyone does as far as I can see. It adds complexity, but I don't think this is necessary.

The prefixes, if available, need to be interpreted by the user anyway. If that wasn't possible, the prefixes would be more than unnecessary bloat of the commit messages.

File: pyuploadtool/changelog/changelog.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    "test":     "Tests",
+    "build":    "Builds",
+    "revert":   "Reverts",
+    "chore":    "Chores",
+    "others":   "Commits"
+}
+
+
+class Changelog:
+    def __init__(self, title=os.getenv("CHANGELOG_TITLE"), commit_link_prefix=None):
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:23]: In Python, one should _never_ have some default value other than `None`. There's more than one reason for this. Also, I don't see where this environment variable is coming from. It's also not documented below.

It'll be easier to explain this if you implement your fallback with a `if ... is None` check, as it gives you more space for explanations.

File: pyuploadtool/changelog/changelog.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        :type title: str
+
+        :param commit_link_prefix: a link prefix, which can be used to show a commit
+        for example
+        commit_link_prefix = https://github.com/$GITHUB_REPOSITORY/commit
+        here, we will add the commit hash to the end.
+        :type commit_link_prefix: str
+        """
+
+        # remove the trailing hash
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:38]: You're stripping slashes, not hashes. Typo?

Also, I (and almost every other Python programmer) can see _what_ you're doing here. But why is this necessary?

A good read: https://standards.mousepawmedia.com/csi.html

File: pyuploadtool/changelog/changelog.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        self.commit_link_prefix = commit_link_prefix.rstrip("/")
+        self.title = title
+
+        self._data = dict()
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:42]: Why are you storing data in a `dict`?

It seems like a separate class for commits is more appropriate, and easier to use. And it's necessary to fulfill the SoC requirements.

File: pyuploadtool/changelog/changelog.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            self._data[spec] = list()
+
+    def push(self, commit: dict):
+        """
+        Adds a commit to the changelog and aligns each commit
+        based on their category. See CONVENTIONAL_COMMIT_SPEC
+        :return: The classification of the commit == CONVENTIONAL_COMMIT_SPEC.keys()
+        :rtype: str
+        """
+        message = commit.pop("message")
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:54]: You should insert more blank lines IMO. I insert blank lines after every docstring to separate it visually from the actual code. Also, before and after loops, I like to have blank lines for the same reason.

File: pyuploadtool/changelog/changelog.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                return spec
+
+        # it did not fit into any proper category, lets push to others
+        self._data["others"].append({
+            "message": message.strip(),
+            **commit
+        })
+        return "others"
+
+    def get_changelog(self):
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:76]: I think it's a bit unfortunate that this method is called `get_changelog`. It doesn't return "the changelog". It renders the changelog in Markdown and returns a string.

We might want to implement a separate `ChangelogFormatter` system of classes to be able to render different flavors (e.g., a pure text version for the WebDAV uploader, Markdown for GitHub releases, ...).

But for now, I'd rather keep it simple. But you should at least rename the method. I suggest something like `to_markdown()` or `render_to_markdown()`.

**Review by TheAssassin — CHANGES_REQUESTED** [review]

File: pyuploadtool/changelog/changelog.py (hunk 3 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog.py:7]: > There's no reason to process any kind of "commit prefix". I don't use them, hardly anyone does as far as I can see. It adds complexity, but I don't think this is necessary.

Well, I use them, and that's one of the only blocker of not using `pyuploadtool` in my projects. I will try to show some projects using it too.

> 
> The prefixes, if available, need to be interpreted by the user anyway. If that wasn't possible, the prefixes would be more than unnecessary bloat of the commit messages.

Hmm, ok... I will still stick with this commit style.

File: pyuploadtool/changelog/changelog.py (hunk 4 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog.py:23]: ok, will fix right away

File: pyuploadtool/changelog/changelog.py (hunk 6 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog.py:42]: What do you mean by SoC? The best I can relate.. is System on a Chip

File: pyuploadtool/changelog/changelog.py (hunk 3 above)
Comment (TheAssassin) [reply pyuploadtool/changelog/changelog.py:7]: You can, and I have not often rejected commits because of such prefixes. The point, however, is that there is no standard or anything similar around these prefixes. _You_ may use this setup, but it's not applicable to the general public. Other people use other abbreviations, or even don't use this at all.

This "expanding" of these abbreviations is therefore IMO nothing that should be done in a tool like this one. It doesn't add any important functionality (the abbreviations can already be interpreted by the user, and git or GitHub don't expand them either). However, it adds complexity to the code base, and must be maintained. There's even a small chance that such guesswork could break other people's Git commit titles.

File: pyuploadtool/releases_hosting_provider/github_releases.py (hunk 1 above)
Comment (srevinsaju) [reply pyuploadtool/releases_hosting_provider/github_releases.py:82]: I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional. 
I do not understand the concept of optional metadata, in this case
* we are not forcing all the users to accept the changelog we generated
* this changelog feature, is currently bound to a single infrastructure.. GitHub Releases

Until we add the support for other platforms, I prefer to keep as is.

File: pyuploadtool/changelog/changelog.py (hunk 6 above)
Comment (TheAssassin) [reply pyuploadtool/changelog/changelog.py:42]: Separation of concerns, a design principle in software architecture.

File: pyuploadtool/changelog/changelog.py (hunk 3 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog.py:7]: I do not expand the prefixes. A trial run would reveal that. I have just replicated this, which is how a changelog'd release which follows commit spec should ideally look like:

https://github.com/srevinsaju/guiscrcpy/releases/tag/v4.7.4

File: pyuploadtool/releases_hosting_provider/github_releases.py (hunk 1 above)
Comment (TheAssassin) [reply pyuploadtool/releases_hosting_provider/github_releases.py]: > I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional.
I do not understand the concept of optional metadata, in this case

How do _you_ define metadata? Metadata here is defined as "some kind of data that can be used to enhance the uploading of releases". Of course changelogs are a sort of metadata.

> I prefer to keep metadata static, and not optional.

It _is_ static (as the commit data do not change once they're calculated). I don't know how to interpret optional. To me, all metadata _can_ be used by release hosting providers. Some is used, some isn't, some is mandatory.

> this changelog feature, is currently bound to a single infrastructure.. GitHub Releases

Adding it to the metadata doesn't mean a release hosting provider _must_ use it. As explained, metadata _may_ be processed. The idea is to gather as much as possible, then present it to the hosting providers, which _can_ but do not have to make any use of it. It's up to the release hosting provider to use any of the data. That adds a _lot_ of flexibility to the code.

> Until we add the support for other platforms, I prefer to keep as is.

This version goes against the underlying architecture of the software, unfortunately.

Comment (srevinsaju) [reply pyuploadtool/releases_hosting_provider/github_releases.py]: > > I feel that making changelog as a part of metadata is redundant. I prefer to keep metadata static, and not optional.
> > I do not understand the concept of optional metadata, in this case
> 
> How do _you_ define metadata? Metadata here is defined as "some kind of data that can be used to enhance the uploading of releases". Of course changelogs are a sort of metadata.
> 
> > I prefer to keep metadata static, and not optional.
> 
> It _is_ static (as the commit data do not change once they're calculated). I don't know how to interpret optional. To me, all metadata _can_ be used by release hosting providers. Some is used, some isn't, some is mandatory.
> 

The  question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit

> > this changelog feature, is currently bound to a single infrastructure.. GitHub Releases
> 
> Adding it to the metadata doesn't mean a release hosting provider _must_ use it. As explained, metadata _may_ be processed. The idea is to gather as much as possible, then present it to the hosting providers, which _can_ but do not have to make any use of it. It's up to the release hosting provider to use any of the data. That adds a _lot_ of flexibility to the code.
> 
> > Until we add the support for other platforms, I prefer to keep as is.
> 
> This version goes against the architecture of the software, unfortunately.

Ok.. temporily, I will add `metadata.changelog` which would be `None` until the user requests changelog. I hope that's fine.

File: pyuploadtool/changelog/changelog.py (hunk 3 above)
Comment (TheAssassin) [reply pyuploadtool/changelog/changelog.py:7]: Right. I mixed this with some other project. You use it as categorization.

The criticism however stays the same. This is nothing that should be enabled by default. If you insist on this "feature", you should make it opt-in IMO.

I think you should create some `CommitSpecChangelog` which subclasses a basic `Changelog`. A factory method (e.g., in a new class) could then decide which one to instantiate. This changelog could then be inserted into the `ReleaseMetadata`. As all types of changelogs share the same interface, the release hosting providers can then, _if_ they make use of the changelog, get a formatted string, and use that to extend the release description, etc.

File: pyuploadtool/changelog/github.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    def _get_github_api_repos(self, arg):
+        """
+        Helper function to aid github api calls.
+
+        :param arg: https://docs.github.com/en/free-pro-team@latest/rest/reference/repos
+        :type arg: str
+        :return: JSON response
+        :rtype: dict
+        """
+        api_endpoint = f"https://api.github.com/repos/{self.github_repository}/{arg}"
```
Comment (TheAssassin) [inline pyuploadtool/changelog/github.py:39]: Why not use the `github` package? What you do here is more complex than letting that package handle everything. Plus, the package is used everywhere else.

File: pyuploadtool/releases_hosting_provider/github_releases.py (hunk 1 above)
Comment (TheAssassin) [reply pyuploadtool/releases_hosting_provider/github_releases.py]: > Until we add the support for other platforms, I prefer to keep as is.

This is the wrong approach. This would be duplicated in any other release hosting provider. In fact, I intend to extend the WebDAV provider with this functionality as well. Do you think it is a good idea to duplicate the code?

The reason I implemented the software architecture as it is is to avoid such scenarios. The code simply doesn't belong here.

> The question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit

You won't run into rate limits easily on GitHub actions. Check how the `github` package is used everywhere. It's using the `$GITHUB_TOKEN` environment variable. These tokens are very unlike to run into rate limits.

> Ok.. temporily, I will add metadata.changelog which would be None until the user requests changelog. I hope that's fine.

That's how the metadata object works. I should probably document that system a bit better. Why would this be temporarily, though?

Comment (srevinsaju) [reply pyuploadtool/releases_hosting_provider/github_releases.py]: > > Until we add the support for other platforms, I prefer to keep as is.
> 
> This is the wrong approach. This would be duplicated in any other release hosting provider. In fact, I intend to extend the WebDAV provider with this functionality as well. Do you think it is a good idea to duplicate the code?
> 
> The reason I implemented the software architecture as it is is to avoid such scenarios. The code simply doesn't belong here.

Ah I see
> 
> > The question is.. the need for calculation. Yes, as you said, its static. I purposefully made it optional to prevent running into GitHub API Ratelimits. For each tag, generating Changelog metadata uses 3 API calls. For an unauthorized API call, the tool would fail due to the redundant API calls and GitHub API's ratelimit
> 
> You won't run into rate limits easily on GitHub actions. Check how the `github` package is used everywhere. It's using the `$GITHUB_TOKEN` environment variable. These tokens are very unlike to run into rate limits.
> 
> > Ok.. temporily, I will add metadata.changelog which would be None until the user requests changelog. I hope that's fine.
> 
> That's how the metadata object works. I should probably document that system a bit better. Why would this be temporarily, though?

Oh, I thought you wanted me to make another class, which should be of `ReleaseMetadata` type. Now it makes sense.

File: pyuploadtool/changelog/github.py (hunk 9 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/github.py]: Fixed and rebased to https://github.com/TheAssassin/pyuploadtool/pull/3/commits/1e62a479f80334d66ddb075d62e25cf6abd5f900

File: pyuploadtool/changelog/changelog.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        """
+        Adds a commit to the changelog
+        :return: The classification of the commit = other
+        :rtype: str
+        """
+        self._data["others"].append(commit)
+        return "others"
+
+    @property
+    def changelog(self):
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:39]: This docstring seems incomplete.

File: pyuploadtool/changelog/changelog.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+    def __getitem__(self, item):
+        return self._data[item]
+
+    @property
+    def structure(self):
+        """
+        Minimal structure of a changelog. All commits
+        would be classified as others
+        :return:
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog.py:24]: This docstring seems incomplete.

File: pyuploadtool/changelog/changelog_spec.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import re
+
+from .changelog import Changelog
+from .commit import ChangelogCommit
+
+
+class ConventionalCommitChangelog(Changelog):
+    @property
+    def structure(self) -> dict:
```
Comment (TheAssassin) [inline pyuploadtool/changelog/changelog_spec.py:9]: This should be a static or class method, not a property. It's not object-specific.

File: pyuploadtool/changelog/commit.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
+from github.Commit import Commit
+
+from .author import Author
+
+
+class ChangelogCommit:
```
Comment (TheAssassin) [inline pyuploadtool/changelog/commit.py:6]: This likely wants to be a
```python
class ChangelogCommit(NamedTuple):
    author: Author
    message: str
    sha: str
```

You don't need to "write-protect" attributes in Python, really.

File: pyuploadtool/changelog/factory/base.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        """
+        Get the corresponding changelog generator from the environment
+        if it is not supplied.
+        :return:
+        :rtype:
+        """
+        if self.changelog_type is None:
+            self.changelog_type = ChangelogType.from_environment()
+
+        _generator = SUPPORTED_CHANGELOG_TYPES.get(self.changelog_type)
```
Comment (TheAssassin) [inline pyuploadtool/changelog/factory/base.py:28]: The leading `_` doesn't add anything here, you could remove it.

File: pyuploadtool/changelog/factory/github.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+        """
+        Gets the latest release by semver, like v8.0.1, v4.5.9, if not
+        Fallback to continuous releases, like 'continuous', 'stable', 'nightly'
+
+        :return: the tag name of the latest release, and the date on which it was created
+        :rtype: GitRelease
+        """
+
+        releases = self.repository.get_releases()
+        _latest_release = None
```
Comment (TheAssassin) [inline pyuploadtool/changelog/factory/github.py:34]: Why exactly are you using underscore prefixes for local variables?

File: pyuploadtool/changelog/parsers/parser.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        changelog: Changelog,
+        title=None,
+        commit_link_prefix=None,
+    ):
+        """
+        Generates a changelog by arranging the commits according
+        to the Conventional Commit Spec
+
+        :param title: the title of the release, generally, the tag name
+        :type title: str
```
Comment (TheAssassin) [inline pyuploadtool/changelog/parsers/parser.py:16]: You can use type annotations instead:
```python3
def __init__(self, ..., title: str = None, ...)

File: pyuploadtool/changelog/types.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+class ChangelogType(Enum):
+    # default
+    STANDARD = 0
+
+    # follows the Conventional Commit Spec
+    CONVENTIONAL = 1
+
+    @staticmethod
+    def from_environment():
+        _type = os.getenv("CHANGELOG_TYPE")
```
Comment (TheAssassin) [inline pyuploadtool/changelog/types.py:14]: Same question: why are you using an underscore prefix?

File: pyuploadtool/releases_hosting_provider/github_releases.py [hunk 18]
```diff
@@ -77,6 +78,19 @@ def create_release(self, metadata: ReleaseMetadata, artifacts):
 
         message = f"Build log: {metadata.build_log_url}"
 
+        if os.getenv("CHANGELOG_GENERATE", "").lower() == "true":
+
```
Comment (TheAssassin) [inline pyuploadtool/releases_hosting_provider/github_releases.py:82]: This line seems superfluous.

File: pyuploadtool/releases_hosting_provider/github_releases.py [hunk 19]
```diff
@@ -77,6 +78,19 @@ def create_release(self, metadata: ReleaseMetadata, artifacts):
...
+            github_changelog = GitHubChangelogFactory(github_client=self.github_client, metadata=metadata)
+            metadata.changelog = github_changelog.get_changelog()
```
Comment (TheAssassin) [inline pyuploadtool/releases_hosting_provider/github_releases.py:84]: What's the point of setting a changelog attribute at this point when you're not using it anyway?

The changelog should be generated in the GitHub actions build system. This side of the tool should then only render the given metadata. It should not add new information. The changelog might be used in other release hosting providers, too, but this would then require this one to run first.

Think: once the build system is done, metadata should not change any more.

**Review by TheAssassin — CHANGES_REQUESTED** [review]
Looks quite good already. I found a workflow issue that you need to resolve, as well as a few minor issues and questions. The architecture looks okay.

I think the changelog generation might be implemented using the strategy pattern, too, but on the other hand, I'd rather like it to be all fetched and prepared in the build system side, then passing the result to the various release hosting providers.

File: pyuploadtool/changelog/commit.py (hunk 13 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/commit.py]: Oh! That's an awesome trick. Thanks!

File: pyuploadtool/changelog/changelog_spec.py (hunk 12 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog_spec.py]: Agreed, yes.

File: pyuploadtool/changelog/types.py (hunk 17 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/types.py]: I am not sure why :man_facepalming: :shrug: lol.

File: pyuploadtool/releases_hosting_provider/github_releases.py (hunk 19 above)
Comment (srevinsaju) [reply pyuploadtool/releases_hosting_provider/github_releases.py]: Ok, I have improved this part a bit more, by not editing the `metadata` in the release section. But however, I still haven't moved the changelog generation part to `build_system` because, 
* `GitHubChangelogFactory` requires a `GitHub` client instance. 
* The generation of changelog is optional, and will only happen if the `metadata.release_description` is not `None`.

I can move the `GitHubChangelogFactory` outside the `release_hosting_provider` but then, that will initialize the `GitHub` client again. Shall I proceed with that?

File: pyuploadtool/changelog/parsers/markdown.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
+from .parser import ChangelogParser
+
+
+class MarkdownChangelogParser(ChangelogParser):
+    def render_to_markdown(self):
+        """
+        Get Markdown MarkdownChangelogParser
```
Comment (TheAssassin) [inline pyuploadtool/changelog/parsers/markdown.py:7]: Pretty broken docstring. The fields below are not even filled in.

**Review by TheAssassin — APPROVED** [review]
Looks really good now. I just found a usability issue and some dead docstrings now. But otherwise, it's pretty much how I'd had integrated it. Good job!

**Review by TheAssassin — CHANGES_REQUESTED** [review]
Fail: meant to request changes.

File: pyuploadtool/changelog/parsers/markdown.py (hunk 20 above)
Comment (srevinsaju) [reply pyuploadtool/changelog/parsers/markdown.py]: Fixed in https://github.com/TheAssassin/pyuploadtool/pull/3/commits/db0952b6011c414d4a4b8e671cd3c8412aad04cf

File: pyuploadtool/changelog/changelog.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    def structure() -> dict:
+        """
+        Returns a dictionary with a minimal structure of a changelog.
+        All commits would be classified as others by default.
+        :return:
+        :rtype:
+        """
+        return {"others": "Commits"}
+
+    def push(self, commit: ChangelogCommit):
```
[the remark this answers is not in the document]
Comment (srevinsaju) [reply pyuploadtool/changelog/changelog.py]: Agreed, you are right. Looks more generic. 

Fixed in https://github.com/TheAssassin/pyuploadtool/pull/3/commits/6eb611fabd46677b6c75beff5c4947bb72b7b2c4

File: README.md [hunk 22]
```diff
@@ -69,3 +69,16 @@ You can upload to any WebDAV server which supports `PUT` operations. The followi
...
 **Note:** Secrets must not be stored inside the repository, nor be visible to end users. You need to store them securely, ideally using the credentials storage your build system provides (on GitHub actions, there's *Secrets*, for instance).
+
+
+## Changelog Generation
+`pyuploadtool` support Changelog generation, which is optional, and can be enabled with the `$CHANGELOG_GENERATE` environment variable.
+```bash
+CHANGELOG_GENERATE=true ./pyuploadtool*.AppImage
+```
+### Changelog Types
+`pyuploadtool` currently supports two types of Changelogs.
```
[the remark this answers is not in the document]
Comment (srevinsaju) [reply README.md]: > Wouldn't it be easier to just have one variable for everything? How big are the chances that CHANGELOG_TYPE is set without the other variable?

Yea, that's good. And that also helps by not hardcoding a default for commit style.

> Also, I'd like to avoid numeric values, and I'd also go for lowercase values and uppercase keys.

Fixed both of these in https://github.com/TheAssassin/pyuploadtool/pull/3/commits/50455a8328ffaa4a822819c1a6a8574c70887d5a and https://github.com/TheAssassin/pyuploadtool/pull/3/commits/6e18de8e35b1bf3f012b1251c74fd071a4bb7328

**Comment by TheAssassin** [comment]
Please rebase the branch and squash/"fixup" all corrections to commits previously made in this PR. The amount of commits is quite high.
